### PR TITLE
Crew monitoring rework part 1: crew pinpointers

### PR DESCRIPTION
**Pull request by pubby**
Crew monitoring consoles have been in a laggy, client-crashing state for a long time. Nobody has bothered to fix them, and it's safe to say nobody will bother fixing them in the foreseeable future. Maybe it's time we rewrite them.

This PR contains an alternative intended to replace one part of the console's functionality: crew pinpointers. A crew pinpointer is just a pinpointer that tracks maxed suit sensors. Crewmembers with their sensors below max can't be tracked.

If this gets merged, the next step will be to replace the consoles with a slimmed-down manifest that lacks a minimap. The consoles will show vitals, but position will be text-only. Use pinpointers to track crewmember position exactly.

Two pinpointers start in each full-size medical vendor. One starts in the detective's cabinet. 

They are painted blue.

:cl: Pubby
add: Crew-tracking pinpointers to replace laggy crew monitoring console functionality
add: Crew-tracking pinpointers in medical vendors and the detective's office
/:cl:

If anyone wants to argue balance before thinking: These provide less information than handheld crew monitors. If anything, this rework is a nerf.

### CODE REVIEW

**Comment by optimumtact** [comment]
I assume the plan for the monitor console is to move it away from html_interface to either tgui or the old dat standard output?

**Comment by pubby** [comment]
That's the plan, operative

**Comment by optimumtact** [comment]
no but what one

**Comment by pubby** [comment]
I was gonna do the old dat standard output unless you guys make me do tgui

I don't have faith in tgui

**Comment by optimumtact** [comment]
nobody does now, replacements are being sought

**Comment by FantasticFwoosh** [comment]
Its simple to destroy said pinpointers, so would you propose that you acquire more from a cargo crate OR RnD?

- The forensic scanner as a example are hard to get hold of once the originals are taken/destroyed and stop people from critically doing their jobs.

File: code/game/objects/items/pinpointer.dm [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+	throw_range = 7
+	materials = list(MAT_METAL = 500, MAT_GLASS = 250)
+	resistance_flags = INDESTRUCTIBLE | LAVA_PROOF | FIRE_PROOF | ACID_PROOF
+	var/active = FALSE
+	var/atom/movable/target = null //The thing we're searching for
+	var/minimum_range = 0 //at what range the pinpointer declares you to be at your destination
+	var/alert = FALSE // TRUE to display things more seriously
+
+/obj/item/pinpointer/New()
+	..()
```
Comment (Cyberboss) [inline code/game/objects/items/pinpointer.dm]: Initialize and . = ..()

File: code/game/objects/items/pinpointer.dm [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+	w_class = WEIGHT_CLASS_SMALL
+	item_state = "electronic"
+	lefthand_file = 'icons/mob/inhands/misc/devices_lefthand.dmi'
+	righthand_file = 'icons/mob/inhands/misc/devices_righthand.dmi'
+	throw_speed = 3
+	throw_range = 7
+	materials = list(MAT_METAL = 500, MAT_GLASS = 250)
+	resistance_flags = INDESTRUCTIBLE | LAVA_PROOF | FIRE_PROOF | ACID_PROOF
+	var/active = FALSE
+	var/atom/movable/target = null //The thing we're searching for
```
Comment (Cyberboss) [inline code/game/objects/items/pinpointer.dm]: remove the ` = null`

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by calzilla1** [comment]
Add one to the wardens office

**Comment by BeeSting12** [comment]
These should be made by RND also. One should be found in the CMO's cabinet and warden's office. 

Can you select which crewmember you're looking for?

**Comment by kevinz000** [comment]
:-1: so fix the lag by removing the minimap and making it text position. Why nerf? If you want someone to be irretrievable either space them or take their suit off, not hard.

**Comment by pubby** [comment]
That's the plan, mekhi, but i did not make it clear

also I'm gonna be out sick for a day or three so bear with me on the changes

**Comment by Cyberboss** [comment]
@pubby do changes

**Comment by pubby** [comment]
@cyberboss changes done

I've added them to the CMO and warden. You can already order more from cargo by buying a NanoMed refill pack.

**Comment by Cyberboss** [comment]
Confrictur

**Comment by pubby** [comment]
:unconfrict:

File: code/game/objects/items/pinpointer.dm [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+/obj/item/pinpointer/Destroy()
+	STOP_PROCESSING(SSfastprocess, src)
+	GLOB.pinpointer_list -= src
+	return ..()
+
+/obj/item/pinpointer/attack_self(mob/living/user)
+	active = !active
+	user.visible_message("<span class='notice'>[user] [active ? "" : "de"]activates their pinpointer.</span>", "<span class='notice'>You [active ? "" : "de"]activate your pinpointer.</span>")
+	playsound(user, 'sound/items/screwdriver2.ogg', 50, 1)
```
Comment (Cyberboss) [inline code/game/objects/items/pinpointer.dm]: user -> src

File: code/game/objects/items/pinpointer.dm [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+		START_PROCESSING(SSfastprocess, src)
+	else
+		target = null
+		STOP_PROCESSING(SSfastprocess, src)
+	update_pointer_overlay()
+
+/obj/item/pinpointer/process()
+	if(!active)
+		STOP_PROCESSING(SSfastprocess, src)
+		return
```
Comment (Cyberboss) [inline code/game/objects/items/pinpointer.dm]: return PROCESS_KILL instead of STOP_PROCESSING

File: code/game/objects/items/pinpointer.dm [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+	scan_for_target()
+	update_pointer_overlay()
+
+/obj/item/pinpointer/proc/scan_for_target()
+	return
+
+/obj/item/pinpointer/proc/update_pointer_overlay()
```
Comment (Cyberboss) [inline code/game/objects/items/pinpointer.dm]: Should just be update_icon() no?

File: code/game/objects/items/pinpointer.dm [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+		var/turf/there = get_turf(H)
+		return (H.z != 0 || (there && there.z == H.z))
+
+	return FALSE
+
+/obj/item/pinpointer/crew/attack_self(mob/living/user)
+	if(active)
+		active = FALSE
+		user.visible_message("<span class='notice'>[user] deactivates their pinpointer.</span>", "<span class='notice'>You deactivate your pinpointer.</span>")
+		playsound(user, 'sound/items/screwdriver2.ogg', 50, 1)
```
Comment (Cyberboss) [inline code/game/objects/items/pinpointer.dm]: user -> src

File: code/game/objects/items/pinpointer.dm [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+		return
+
+	var/list/name_counts = list()
+	var/list/names = list()
+
+	for(var/mob/living/carbon/human/H in GLOB.mob_list)
+		if(!trackable(H))
+			continue
+
+		var/name = "Unknown"
```
Comment (Cyberboss) [inline code/game/objects/items/pinpointer.dm]: name this var something else

File: code/game/objects/items/pinpointer.dm [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+			name = text("[] ([])", name, name_counts[name])
+		names[name] = H
+		name_counts[name] = 1
+
+	if(!names.len)
+		user.visible_message("<span class='notice'>[user]'s pinpointer fails to detect a signal.</span>", "<span class='notice'>Your pinpointer fails to detect a signal.</span>")
+		return
+
+	var/A = input(user, "Person to track", "Pinpoint") in names
+	if(!src || QDELETED(src) || !user || !user.is_holding(src) || user.incapacitated() || !A)
```
Comment (Cyberboss) [inline code/game/objects/items/pinpointer.dm]: `if(!A || QDELETED(src) || QDELETED(user) || user.incapacitated() || !user.is_holding(src))`

File: code/game/objects/items/pinpointer.dm [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+		return
+
+	target = names[A]
+	active = TRUE
+	user.visible_message("<span class='notice'>[user] activates their pinpointer.</span>", "<span class='notice'>You activate your pinpointer.</span>")
+	playsound(user, 'sound/items/screwdriver2.ogg', 50, 1)
```
Comment (Cyberboss) [inline code/game/objects/items/pinpointer.dm]: user -> src

File: code/game/objects/items/pinpointer.dm [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+	START_PROCESSING(SSfastprocess, src)
+	update_pointer_overlay()
+
+/obj/item/pinpointer/crew/scan_for_target()
+	if(target)
+		if(ishuman(target))
+			var/mob/living/carbon/human/H = target
+			if(!trackable(H))
+				target = null
+	if(!target)
```
Comment (Cyberboss) [inline code/game/objects/items/pinpointer.dm]: should just be `else`

File: code/game/objects/items/pinpointer.dm [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+		active = FALSE
+
+/obj/item/pinpointer/process()
+	if(!active)
+		STOP_PROCESSING(SSfastprocess, src)
+		return
```
Comment (Cyberboss) [inline code/game/objects/items/pinpointer.dm]: PROCESS_KILL

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Review by Cyberboss — APPROVED** [review]

**Pull request merged**